enhancements/ingress/logging-headers.md

```diff
@@ -0,0 +1,260 @@
+- "enhancements/ingress/logging-api.md"
+replaces:
+superseded-by:
+---
+
+# Ingress Header Logging
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement extends the IngressController logging API to allow the user to
+configure capture of HTTP request and response headers in access logs.  By
+default, when access logging is enabled, the IngressController logs the HTTP
+request line and response code but no headers.  Using the new API, users may
+specify a list of arbitrary HTTP request headers and a list of response headers
+that should be captured and logged when access logging is enabled.
+
+## Motivation
+
+Certain headers may need to be logged for compliance reasons or other reasons.
+In OpenShift 3, cluster administrators could use a custom HAProxy configuration
+template to customize logging, but using a custom template is not permitted in
+OpenShift 4.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP request and response headers that should be captured.
+2. Enable the cluster administrator to specify a custom HTTP log format string that references the captured headers.
+
+### Non-Goal
+
+1. Providing control on a per-frontend or per-route basis.
+
+## Proposal
+
+The IngressController API is extended by adding an optional `HTTPCaptureHeaders`
+field with type `*IngressControllerCaptureHTTPHeaders` to `AccessLogging`:
+
+```go
+type AccessLogging struct {
+	// ...
+
+	// httpCaptureHeaders defines HTTP headers that should be captured in
+	// access logs.  If this field is empty, no headers are captured.
+	//
+	// +optional
+	HTTPCaptureHeaders *IngressControllerCaptureHTTPHeaders `json:"httpCaptureHeaders,omitempty"`
+}
+```
+
+The `IngressControllerCaptureHTTPHeaders` type has a `Request` field and a
+`Response` field, both of type `[]IngressControllerCaptureHTTPHeader`, for
+specifying headers to capture:
+
+```go
+// IngressControllerCaptureHTTPHeaders specifies which HTTP headers the
+// IngressController captures.
+type IngressControllerCaptureHTTPHeaders struct {
+	// request specifies which HTTP request headers to capture.
+	//
+	// If this field is empty, no request headers are captured.
+	Request []IngressControllerCaptureHTTPHeader `json:"request,omitempty"`
+
+	// response specifies which HTTP response headers to capture.
+	//
+	// If this field is empty, no response headers are captured.
+	Response []IngressControllerCaptureHTTPHeader `json:"response,omitempty"`
+}
+```
+
+The `IngressControllerCaptureHTTPHeader` type has fields for specifying a header
+name and maximum length for a captured header value (header values that exceed
+the maximum length are truncated to that length):
```

Comment:
> (header values that exceed
the maximum length are truncated to that length)

What about maximum length values that exceed the length of the header value? Is only the applicable header read?

Reply:
If you specify a max length of 5:

```yaml
spec:
  logging:
    access:
      httpCaptureHeaders:
        request:
        - name: X-My-Header
          maxLength: 5
```

and an HTTP request has the specified header with a value that is shorter than 5 characters:

```http
GET / HTTP/1.1
X-My-Header: foo
```

then the entire value (`foo`) is captured.  if the value is longer than 5 characters:

```http
GET / HTTP/1.1
X-My-Header: foobar
```

then the value is truncated (the logs will show `{fooba}`).  Is that what you mean?

Reply:
The (`foo`) example was what I meant. You answered my question! From your explanation, setting a `maxLength` to be greater than the expected length of the header is not harmful.